lib/ansible/galaxy/collection/galaxy_api_proxy.py

```diff
@@ -84,28 +88,44 @@ def _get_collection_versions(self, requirement: Requirement) -> t.Iterator[tuple
         if not found_api and last_error is not None:
             raise last_error
 
+    def _supports_ansible_version(self, requires_ansible : str | None) -> bool:
+        """Test if a requires_ansible value is compatible with the ansible-core version."""
+        if requires_ansible is None:
+            return C.COLLECTIONS_ON_UNDOCUMENTED_ANSIBLE_VERSION == 'ignore'
+        if C.COLLECTIONS_ON_ANSIBLE_VERSION_MISMATCH == 'ignore':
+            return True
+        return _does_collection_support_ansible_version(requires_ansible, ansible_version)
+
     def get_collection_versions(self, requirement: Requirement) -> t.Iterable[tuple[str, GalaxyAPI]]:
         """Get a set of unique versions for FQCN on Galaxy servers."""
+        versions = 0
+        compatible_versions = []
         if requirement.is_concrete_artifact:
-            return {
-                (
-                    self._concrete_art_mgr.
-                    get_direct_collection_version(requirement),
-                    requirement.src,
-                ),
-            }
+            versions = 1
+            version = self._concrete_art_mgr.get_direct_collection_version(requirement)
+            requires_ansible = self._concrete_art_mgr.get_requires_ansible(requirement)
+            if self._supports_ansible_version(requires_ansible):
+                compatible_versions.append((version, requirement.src))
+        else:
+            skipped_versions: dict[GalaxyAPI, list[str]] = {}
+            for api, version in self._get_collection_versions(requirement):
+                versions += 1
+                requires_ansible = (api.requires_ansible.get(requirement.fqcn) or {}).get(version)
+                if self._supports_ansible_version(requires_ansible):
```

Comment:
Could you explain why can't this be in `find_matches()` in providers?

Reply:
This doesn't feel right.. The API proxy abstraction was meant to be a thin layer for accessing different collection sources. But the more I read this, the more I'm convinced that this belongs more to the resolver mechanics.

Reply:
That might be, but my concern was that if things start leaking across layers of abstraction, this would break the structure I've been envisioning all those years ago.

Anyway, since you've moved away from this, I'm marking this thread as resolved.

Reply:
I'm glad you consider your concern resolved despite our difference of opinion on that. This kind of thing is why it's more useful to bikeshed implementation details after reviewing functionality since it can make that kind of thing moot.